File: plugins/e2/Zen Coding.tmbundle/Support/zen_editor.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

'''
High-level editor interface that communicates with e2 editor.
 
@author Sergey Chikuyonok (serge.che@gmail.com)
@link http://chikuyonok.ru
'''
import os
import sys
import zencoding
import subprocess
import re
import hessian_ipc

class ZenEditor():
	def __init__(self, context=None):
		self.set_context(context)

	def set_context(self, context=None):
		"""
		Setup underlying editor context. You should call this method 
		<code>before</code> using any Zen Coding action.
		@param context: context object
		"""
		self.context = hessian_ipc.Hessian()
		self.editor = self.context.GetActiveEditor()
		
		if self.context.IsSoftTabs():
			indent = ' ' * self.context.GetTabWidth()
			zencoding.utils.set_variable('indentation', indent)
		
	def get_selection_range(self):
		"""
		Returns character indexes of selected text
		@return: list of start and end indexes
		"""
		sels = self.editor.GetSelections()
		if sels:
			sel = sels[0]
			return sel[0], sel[1]
		else:
			pos = self.editor.GetPos()
			return pos, pos
	
	
	def create_selection(self, start, end=None):
		"""
		Creates selection from <code>start</code> to <code>end</code> character
		indexes. If <code>end</code> is ommited, this method should place caret 
		and <code>start</code> index
		"""
		if end is not None:
			self.editor.Select(start, end)
		else:
		    self.editor.SetPos(start)
	
	def get_current_line_range(self):
		"""
		Returns current line's start and end indexes
		@return: list of start and end indexes
		@example
		start, end = zen_editor.get_current_line_range();
		print('%s, %s' % (start, end))
		"""
		line_id = self.editor.GetCurrentLine()
		line_range = self.editor.GetLineRange(line_id)
		return line_range[0], line_range[1]
	
	def get_caret_pos(self):
		""" Returns current caret position """
		return self.editor.GetPos()
	
	def set_caret_pos(self, pos):
		"""
		Set new caret position
		@type pos: int
		"""
		self.editor.SetPos(pos)
	
	def get_current_line(self):
		"""
		Returns content of current line
		@return: str
		"""
		line_id = self.editor.GetCurrentLine()
		return self.editor.GetLineText(line_id)
	
	def replace_content(self, value, start=None, end=None, no_indent=False):
		"""
		Replace editor's content or it's part (from <code>start</code> to 
		<code>end</code> index). If <code>value</code> contains 
		<code>caret_placeholder</code>, the editor will put caret into 
		this position. If you skip <code>start</code> and <code>end</code>
		arguments, the whole target's content will be replaced with 
		<code>value</code>. 
		
		If you pass <code>start</code> argument only,
		the <code>value</code> will be placed at <code>start</code> string 
		index of current content. 
		
		If you pass <code>start</code> and <code>end</code> arguments,
		the corresponding substring of current target's content will be 
		replaced with <code>value</code> 
		@param value: Content you want to paste
		@type value: str
		@param start: Start index of editor's content
		@type start: int
		@param end: End index of editor's content
		@type end: int
		"""
		if start is None: start = 0
		if end is None: end = self.editor.GetLength()
		self.create_selection(start, end)
		
		value = self.add_placeholders(value)
		
		self.editor.InsertSnippet(value)
		
	def log(self, value):
		self.context.Log(value)
		
	
	def get_content(self):
		"""
		Returns editor's content
		@return: str
		"""
		return self.editor.GetText()
	
	def get_syntax(self):
		"""
		Returns current editor's syntax mode
		@return: str
		"""
		scope = self.editor.GetScope()
		cur_scope = scope[1] if len(scope) > 1 else scope[0] 
		default_type = 'html'
		doc_type = None
		try:
			if 'xsl' in cur_scope:
				doc_type = 'xsl'
			else:
				doc_type = re.findall(r'\bcss|xml|haml\b', cur_scope)[-1]
		except:
			doc_type = default_type
		
		if not doc_type: doc_type = default_type
		
		return doc_type
	
	def get_profile_name(self):
		"""
		Returns current output profile name (@see zencoding.utils.setup_profile())
		@return {String}
		"""
		return zencoding.utils.get_variable('profile') or 'xhtml'
	
	def prompt(self, title):
		"""
		Prompt user with CocoaDialog
		@param title: Popup title
		@return: str
		"""
		text = self.editor.Prompt(title)
		if not text:
			return None
		else:
			return text
		
	def add_placeholders(self, text):
		_ix = [0]
		
		def get_ix(m):
			_ix[0] += 1
			return '$%s' % _ix[0]
		
		text = re.sub(r'\$', '\\$', text)
		return re.sub(zencoding.utils.get_caret_placeholder(), get_ix, text)

	def get_selection(self):
		sel_start, sel_end = self.get_selection_range()
		return self.get_content()[sel_start:sel_end]
	
	def get_file_path(self):
		"""
		Returns current editor's file path
		@return: str
		@since: 0.65 
		"""
		return os.getenv('TM_FILEPATH', '').decode('utf-8')

File: plugins/komodo/zencoding/content/js/zen_editor.js
/**
 * High-level editor interface that communicates with underlying editor (like 
 * TinyMCE, CKEditor, etc.) or browser.
 * Basically, you should call <code>zen_editor.setContext(obj)</code> method to
 * set up undelying editor context before using any other method.
 * 
 * This interface is used by <i>zen_actions.js</i> for performing different 
 * actions like <b>Expand abbreviation</b>  
 * 
 * @example
 * var textarea = document.getElemenetsByTagName('textarea')[0];
 * zen_editor.setContext(textarea);
 * //now you are ready to use editor object
 * zen_editor.getSelectionRange();
 * 
 * @author Sergey Chikuyonok (serge.che@gmail.com)
 * @link http://chikuyonok.ru
 */
var zen_editor = {
	context: null,
	scimoz: null,
	
	_bytesToChar: function(bytes) {
		return this.scimoz.getTextRange(0, bytes).length;
	},
	
	_charsToByte: function(chars) {
		var old_pos = this.scimoz.currentPos,
			_c = chars,
			bytes = null;
			
		if (_c) {
			// find correct pos
			do {
				this.scimoz.gotoPos(_c);
				if (this._bytesToChar(this.scimoz.currentPos))
					break;
			} while(_c--);
		}
			
		this.scimoz.gotoPos(_c);
		while (chars > this._bytesToChar(this.scimoz.currentPos)) {
			this.scimoz.charRight();
		}
		
		bytes = this.scimoz.currentPos;
		
		// revert position
		this.scimoz.gotoPos(old_pos);
		
		return bytes;
	},
	
	/**
	 * Setup underlying editor context. You should call this method 
	 * <code>before</code> using any Zen Coding action.
	 * @param {Object} context
	 */
	setContext: function(context) {
		this.context = context;
		this.scimoz = context.scintilla.scimoz;
		
		var indentation = zen_coding.getVariable('indentation');
		if (this.scimoz.useTabs) {
			indentation = '\t';
		} else {
			indentation = zen_coding.repeatString(' ', this.scimoz.indent);
		}
		zen_coding.setVariable('indentation', indentation);
		
		switch (this.scimoz.eOLMode) {
			case 0: // windows
				zen_coding.setNewline('\r\n');
				break;
			case 1: // mac os classic 
				zen_coding.setNewline('\r');
				break;
			default: // unix 
				zen_coding.setNewline('\n');
		}
	},
	
	/**
	 * Returns character indexes of selected text: object with <code>start</code>
	 * and <code>end</code> properties. If there's no selection, should return 
	 * object with <code>start</code> and <code>end</code> properties referring
	 * to current caret position
	 * @return {Object}
	 * @example
	 * var selection = zen_editor.getSelectionRange();
	 * alert(selection.start + ', ' + selection.end); 
	 */
	getSelectionRange: function() {
		return {
			start: this._bytesToChar(this.scimoz.selectionStart),
			end: this._bytesToChar(this.scimoz.selectionEnd)
		};
	},
	
	/**
	 * Creates selection from <code>start</code> to <code>end</code> character
	 * indexes. If <code>end</code> is ommited, this method should place caret 
	 * and <code>start</code> index
	 * @param {Number} start
	 * @param {Number} [end]
	 * @example
	 * zen_editor.createSelection(10, 40);
	 * 
	 * //move caret to 15th character
	 * zen_editor.createSelection(15);
	 */
	createSelection: function(start, end) {
		start = this._charsToByte(start);
		end = this._charsToByte(end);
		this.scimoz.setSel(start, end);
	},
	
	/**
	 * Returns current line's start and end indexes as object with <code>start</code>
	 * and <code>end</code> properties
	 * @return {Object}
	 * @example
	 * var range = zen_editor.getCurrentLineRange();
	 * alert(range.start + ', ' + range.end);
	 */
	getCurrentLineRange: function() {
		var line = this.scimoz.lineFromPosition(this.getCaretPos());
		var result = {
			start: this._bytesToChar(this.scimoz.positionFromLine(line)), 
			end: this._bytesToChar(this.scimoz.getLineEndPosition(line))
		};
		
		return result;
	},
	
	/**
	 * Returns current caret position
	 * @return {Number|null}
	 */
	getCaretPos: function(){
		return this._bytesToChar(this.scimoz.currentPos);
	},
	
	/**
	 * Set new caret position
	 * @param {Number} pos Caret position
	 */
	setCaretPos: function(pos){
		pos = this._charsToByte(pos);
		this.scimoz.currentPos = pos;
		this.scimoz.anchor = pos;
	},
	
	/**
	 * Returns content of current line
	 * @return {String}
	 */
	getCurrentLine: function() {
		var range = this.getCurrentLineRange();
		return this.getContent().substring(range.start, range.end);
	},
	
	/**
	 * Replace editor's content or it's part (from <code>start</code> to 
	 * <code>end</code> index). If <code>value</code> contains 
	 * <code>caret_placeholder</code>, the editor will put caret into 
	 * this position. If you skip <code>start</code> and <code>end</code>
	 * arguments, the whole target's content will be replaced with 
	 * <code>value</code>. 
	 * 
	 * If you pass <code>start</code> argument only,
	 * the <code>value</code> will be placed at <code>start</code> string 
	 * index of current content. 
	 * 
	 * If you pass <code>start</code> and <code>end</code> arguments,
	 * the corresponding substring of current target's content will be 
	 * replaced with <code>value</code>. 
	 * @param {String} value Content you want to paste
	 * @param {Number} [start] Start index of editor's content
	 * @param {Number} [end] End index of editor's content
	 * @param {Boolean} [no_indent] Do not auto indent <code>value</code>
	 */
	replaceContent: function(value, start, end, no_indent) {
		var content = this.getContent(),
			caret_pos = this.getCaretPos(),
			caret_placeholder = zen_coding.getCaretPlaceholder(),
			has_start = typeof(start) !== 'undefined',
			has_end = typeof(end) !== 'undefined';
			
		// indent new value
		if (!no_indent) {
			var line_padding = (this.getCurrentLine().match(/^(\s+)/) || [''])[0];
			value = zen_coding.padString(value, line_padding);
		}
		
		// find new caret position
		var new_pos = value.indexOf(caret_placeholder);
		if (new_pos != -1) {
			caret_pos = (start || 0) + new_pos;
			value = value.split(caret_placeholder).join('');
		} else {
			caret_pos = value.length + (start || 0);
		}
		
		if (!has_start && !has_end) {
			start = 0;
			end = content.length;
		} else if (!has_end) {
			end = start;
		}
		
		this.context.setFocus();
		this.scimoz.beginUndoAction();
		this.scimoz.targetStart = this._charsToByte(start);
		this.scimoz.targetEnd = this._charsToByte(end);
		this.scimoz.replaceTarget(value.length, value);
		this.setCaretPos(caret_pos);
		this.scimoz.endUndoAction();
	},
	
	/**
	 * Returns editor's content
	 * @return {String}
	 */
	getContent: function(){
		return this.scimoz.text;
	},
	
	/**
	 * Returns current editor's syntax mode
	 * @return {String}
	 */
	getSyntax: function(){
		var know_syntaxes = {
			'html': 1,
			'css': 1,
			'xml': 1,
			'xml': 1,
			'haml': 1
		};
		var syntax = this.context.document.language.toLowerCase(),
			caret_pos = this.getCaretPos();
		
		if (!(syntax in know_syntaxes)) {
			syntax = 'html';
		}
		
		if (syntax == 'html') {
			// get the context tag
			var pair = zen_coding.html_matcher.getTags(this.getContent(), caret_pos);
			if (pair && pair[0] && pair[0].type == 'tag' && pair[0].name.toLowerCase() == 'style') {
				// check that we're actually inside the tag
				if (pair[0].end <= caret_pos && pair[1].start >= caret_pos)
					syntax = 'css';
			}
		}
		
		return syntax;
	},
	
	/**
	 * Returns current output profile name (@see zen_coding#setupProfile)
	 * @return {String}
	 */
	getProfileName: function() {
		return 'xhtml';
	},
	
	/**
	 * Ask user to enter something
	 * @param {String} title Dialog title
	 * @return {String} Entered data
	 * @since 0.65
	 */
	prompt: function(title) {
		return ko.dialogs.prompt(title);
	},
	
	/**
	 * Returns current selection
	 * @return {String}
	 * @since 0.65
	 */
	getSelection: function() {
		var sel = getSelectionRange();
		if (sel) {
			try {
				return getContent().substring(sel.start, sel.end);
			} catch(e) {}
		}
		
		return '';
	},
	
	/**
	 * Returns current editor's file path
	 * @return {String}
	 * @since 0.65 
	 */
	getFilePath: function() {
		return ko.views.manager.currentView.document.file.URI;
	},
	
	/**
	 * Returns core Zen Codind object
	 */
	getCore: function() {
		return zen_coding;
	},
	
	/**
	 * Returns Zen Coding resource manager. You can add new snippets and 
	 * abbreviations with this manager, as well as modify ones.<br><br>
	 * 
	 * Zen Coding stores settings in two separate vocabularies: 'system' 
	 * and 'user'. The ultimate solution to add new abbreviations and
	 * snippets is to setup a 'user' vocabulary, like this:
	 * 
	 * @example
	 * var my_settings = {
	 * 	html: {
	 * 		abbreviations: {
	 * 			'tag': '<div class="mytag">'
	 * 		}
	 * 	}
	 * };
	 * zen_editor.getResourceManager().setVocabulary(my_settings, 'user')
	 * 
	 * @see zen_resources.js
	 */
	getResourceManager: function() {
		return zen_resources;
	}
};